cilium/CFP-28427-fqdn-selectors.md

```diff
@@ -0,0 +1,409 @@
+
+Since, in v1.15 and before, the only possible labels are the CIDR labels themselves and the `reserved:kube-apiserver` label, the simple 1:1 mapping between identity and prefix is simple to recreate.
+
+#### Proposal
+
+As always, the numeric identity for a given prefix should, as much as possible, be stable. The reconstruction heuristics will need to change, as a single ID can correspond to multiple prefixes. We can no longer use a simple heuristic to re-create identities.
+
+However, all hope is not lost. The state of the FQDN policy engine is also checkpointed to disk, so we can mostly recreate the set of labels that apply to a given IP. Additionally, we will **checkpoint the local allocator state to disk**, so that the same set of labels receives the same identity.
+
+Checkpointing the local allocator is required so that policy maps (which contain only numeric IDs) remain stable across restarts.
+
+The full procedure is as follows:
+
+1. Read the most recent allocator state. This maps Labels -> Numeric Identity (NID). Insert this in the allocator.
+2. Read the IPCache state. This maps Prefix -> Numeric Identity
+3. For any NIDs in the IPCache and not in the allocator, use the "withold + request" mechanism to request the same NID again
+4. Wait the synchronization grace period (10 minutes)
+5. Release all restored references (identity allocations, placeholder ipcache metadata)
+
+In this proposal, the Label -> NID mapping is now fixed. Thus, for prefixes to have stable NIDs across restarts, they must have the same labels. The agent waits for its data sources to synchronize before computing labels, so this should primarily be the case.
+
+#### Example
+
+The IPCache has three prefixes, each with a corresponding identity.
+
+- `192.168.0.0/24` -> `(cidr:192.168.0.0/24)`
+- `1.1.1.1/32` -> `(fqdn:*, fqdn:one.one.one.one)`
+- `2.2.2.2/32` -> `(fqdn:*)`
+
+If, on agent restart, all prefixes have the same set of labels, then the identity mapping is stable. If, however, the FQDN checkpointing is not exact (which does happen when an IP is near expiration), then `1.1.1.1/32` may change labels to `(fqdn:*)`. However, since the identity for `fqdn:*` is already allowed by policy maps, no traffic to `fqdn:*` is dropped.
+
+Thus, transitioning to label-based, rather than prefix-based allocation should preserve policy stability on restarts.
+
+### Key question: Circuit breakers?
+
+There is currently a circuit-breaker, `--tofqdns-endpoint-max-ip-per-hostname`, which is intended to prevent names such as S3 causing identity churn. In this proposal, large names such as S3 should not allocate more than one or two identities, making this less critical.
+
+However, this is still a security-sensitive area, since we are allocating resources based on external, untrusted packets. We will need a sane circuit-breaker to ensure we do not allocate a large number of identities based on malicious input.
```

Comment:
I think it's fair to say that the trust in the DNS server is also quite critical in general with the ToFQDNs feature, in that if the path to the DNS server or the DNS server itself is somehow compromised, or if the user allows DNS requests to arbitrary destinations and used these as input to the FQDN engine, then these situations could each cause adverse behavior. 

Then there's the 'selector combinations' side of things which is also subject to the network policies configured by the k8s namespace users.

Reply:
Thinking very hard about this, I don't think this proposal significantly changes the security scope or attack surface of the DNS proxy. If an FQDN selector selects an attacker-controlled domain, then the attacker can insert any IPs they want in to policy (waves hand).

Perhaps this is worth a better writeup (@ferozsalam :smile_cat: ).